modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointWorkflow.java

```diff
@@ -396,10 +396,13 @@ private void fillCacheGroupState(CheckpointRecord cpRec) throws IgniteCheckedExc
                     if (partState == LOST)
                         partState = OWNING;
 
+                    assert part.highestAppliedCounter() >= part.updateCounter() :
```

Comment:
I'm not sure we need this assert:
1. PartitionCounter implementation should guarantee that by default, otherwise it's a bug. 
2. `part.highestAppliedCounter()` is a synchronized method, and you invoke it twice here.
3. I have concerns for this expression, is there guarantees that there are no concurrent updates of partition counter between invokation `highestAppliedCounter` and invokation `updateCounter`?

WDYT?

Reply:
> Same. This is checkoint time. Should not be updates.

I checked `PartitionUpdateCounterTrackingImpl#update(long start, long delta)` and it looks like there are some places where it's invoked without acquiring the checkpoint lock, especially when transaction is rollbacked. If I'm correct, then It looks like that such cases don't change the data in partition, but it concurrently updates the counter. Then this assert might fail.

Reply:
There is no assert any more. But isn't it a bug with counter?